tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py

```diff
@@ -395,3 +398,174 @@ def test_invalid(self):
+    # adapted from scipy's analogous tests
+    def _define_cases(self, original, dtype):
+        cases = []
+        cases.append((cp.array(original, dtype=dtype), original))
+        cases.append((cupyx.scipy.sparse.csr_matrix(
+            original, dtype=dtype), original))
+
+        # Test default implementations of _adjoint and _rmatvec, which
+        # refer to each other.
+        def mv(x, dtype):
+            y = original.dot(x)
+            if len(x.shape) == 2:
+                y = y.reshape(-1, 1)
+            return y
+
+        def rmv(x, dtype):
+            return original.T.conj().dot(x)
+
+        class BaseMatlike(interface.LinearOperator):
+            args = ()
+
+            def __init__(self, dtype):
+                self.dtype = cp.dtype(dtype)
+                self.shape = original.shape
+
+            def _matvec(self, x):
+                return mv(x, self.dtype)
+
+        class HasRmatvec(BaseMatlike):
+            args = ()
+
+            def _rmatvec(self, x):
+                return rmv(x, self.dtype)
+
+        class HasAdjoint(BaseMatlike):
+            args = ()
+
+            def _adjoint(self):
+                shape = self.shape[1], self.shape[0]
+                matvec = partial(rmv, dtype=self.dtype)
+                rmatvec = partial(mv, dtype=self.dtype)
+                return interface.LinearOperator(matvec=matvec,
+                                                rmatvec=rmatvec,
+                                                dtype=self.dtype,
+                                                shape=shape)
+
+        class HasRmatmat(HasRmatvec):
+            def _matmat(self, x):
+                return original.dot(x)
+
+            def _rmatmat(self, x):
+                return original.T.conj().dot(x)
+
+        cases.append((HasRmatvec(dtype), original))
+        cases.append((HasAdjoint(dtype), original))
+        cases.append((HasRmatmat(dtype), original))
+        return cases
+
+    def _make_cases(self):
+        self.cases = []
+        original = cp.array([[1., 2., 3.], [4., 5., 6.]])
+        self.cases += self._define_cases(original, cp.float32)
+        self.cases += self._define_cases(original, cp.float64)
+        self.cases += [(interface.aslinearoperator(M).T, A.T)
+                       for M, A in self._define_cases(original.T, cp.float64)]
+        self.cases += [(interface.aslinearoperator(M).H, A.T.conj())
+                       for M, A in self._define_cases(original.T, cp.float64)]
+
+        original = cp.array([[1, 2j, 3j], [4j, 5j, 6]])
+        self.cases += self._define_cases(original, cp.complex_)
+        self.cases += [(interface.aslinearoperator(M).T, A.T)
+                       for M, A in self._define_cases(original.T, cp.complex_)]
+        self.cases += [(interface.aslinearoperator(M).H, A.T.conj())
+                       for M, A in self._define_cases(original.T, cp.complex_)]
+
+    def test_basic(self):
+        for M, A_array in self.cases:
```

Comment:
When the test failed for some bugs, this implementation does not show in which test case failed. Could you use `testing.parameterize`?

Reply:
Could you parameterize `self.cases` by `testing.parameterize`? Just splitting method `test_basic` is also fine.